keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they releate.
+-->
+
+### Risks and Mitigations
+
+- *Risk*: creating another way on how to build packages  
+  *Mitigations*:  
+  - work with the current builder (@sumi) in google
+  - writing super extra great docs so this is the goto way on how to build
+- *Risk*: `nfpm` is not the right tool, it goes away, works differently, ...  
+  *Mitigations*:  
+  - we encapsulate the build of exactly one package, if things change, we can
+    change the internal implementation but keep the external
+    interface
+- *Risk*: We cannot build proper windows packages but community wants/needs them  
+  *Mitigations*:
+  - Talk to people in sig windows, esp. about packaging their components
+
+## Design Details
+
+### Test Plan
+
+- different components will have unit tests as much as possible
+- comparing output debs/rpms to what currently is released
+  - not sure how? by the files they deploy?
+
+### Graduation Criteria
+
+#### From `nil` to `k/k`, Introduce the new mechanism into `k/k`
+
+- Iterate here on the KEP
+- Do PoCs
+- Compare to packages generated with the current release tooling
+- See no significant difference in the generated packages
+- The new mechanism has means to prepare the packages in a way that they can be
+  published via the current release workflow
+
+#### Deprecate all other options to build packages
+
+Deprecation of other packaging mechanisms is done either by a note in the
+README or even in the source code to print a warning when building packages via
+deprecated tools.
+
+Before deprecation the following points should be implemented:
+
+- Make it easy for the release team to switch between the old and the new process
+- Make it easy for the release team to compare two (sets of) packages and
+  assess if they are sufficiently similar
+- Change the release (branch management & patch management) processes to use
+  the new mechanism by default
+
+#### Make this the only available way to build packages
+
+- X (TBD) releases have been cut with the new machanism
+- There was no feedback on bad packages
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+N/A
+
+<!--
+**Note:** *Section not required until targeted at a release.*
+
+If applicable, how will the component be upgraded and downgraded?  Make sure this is in the test plan.
+-->
+
+### Version Skew Strategy
+
+One of the issues we saw is that `k/k` has release branches, while `k/release`
+didn't have the same branch/release strategy.
+This is fixed by adding the new tooling and package building specs to `k/k`
```

Comment:
... Which will then require maintaining the tool across branches too. It might be a better idea to only put the specs in k/k and keep the tool code in k/release if it is significant at all. similarly I would not recommend moving anago into k/k, and some other tooling lives in k/repo-infra.

Reply:
What I thought:
- `anago` will stay in `k/release` - no branching or so
- the orchestrator tool and the packager tool live in `k/k` and change with releases on different branches as needed
- the specs, systemd files, ... all live in `k/k` and change with releases on different branches as needed
- the orchestrators default behaviour, when called with no / the minimum needed set of options, does what we need for the release process
- there should be options for users to change the orchestrators behaviour (e.g. only build platform `$KUBE_BUILD_PLATFORM` build for a specific rev of `k/k`, build from source or pull artefacts from staging bucket, ...)


That means that the probably the call arguments or environment variables and the output the "orchestrator" tool uses need to be defined better, so that the "unversioned" `anago` or package uploader (all in `k/release`) can use the "versioned" tool (in `k/k`) properly. The specs & co, the package configuration, can evolve more freely, as it is versioned the same way as the packager.
Is this fair?

Reply:
my vote goes for the release tooling to change with release branches.
once a release is cut the release tool mechanism is archived and changes need to be backported.
this also ensures that we don't break old releases by introducing a change in the tool.

but on the topic of should it live in k/k i'm tempted to say no, as we should be moving code away from it.
yet, perhaps the release tool is **one** of the things that should live there.

Reply:
We can also do that by pinning to a release tool commit in k/k without actually putting the development of it in k/k, which is I think what I intended to suggest here.

We definitely should freeze all tooling at release time, CI does this as much as possible already.

Reply:
I *think* you *might* end up duplicating the same branching strategy and cherry-picking in the release tooling repo as in `k/k` if there is e.g. a patch release which needs a change to the release tool. I don't want to say that is necessarily a bad thing and there might be some tooling and strategies available from sig-testing, the publishing-bot maybe and other places and teams. I am just wondering if it might be easier, "more natural" and important enough to keep it in `k/k` and therefor in sync with whatever `k/k` is doing.

Another aspect here *might* be: It is maybe easier to do this in `k/k` to make it easy to move forward and evolve with out potentially breaking the release of patches for older k releases?

Lot of not-having-a-strong-opinion, unknowns and ¯\\\_(ツ)\_/¯ from my side here ...